File: src/main/java/com/github/Doomsdayrs/Jikan4java/types/Main/Character/Character.java
package com.github.Doomsdayrs.Jikan4java.types.Main.Character;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.github.Doomsdayrs.Jikan4java.core.Retriever;
import com.github.Doomsdayrs.Jikan4java.types.Support.Pictures.Pictures;
import com.github.Doomsdayrs.Jikan4java.types.Support.Voice_actors;
import org.json.simple.parser.ParseException;

import java.io.IOException;
import java.util.ArrayList;
import java.util.concurrent.CompletableFuture;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 * ====================================================================
 * Jikan4java
 * 28 / October / 2018
 *
 * @author github.com/doomsdayrs
 */
@JsonFormat(shape = JsonFormat.Shape.OBJECT)
public class Character extends Retriever {
    @JsonProperty("request_hash")
    public String request_hash;
    @JsonProperty("request_cached")
    public boolean request_cached;
    @JsonProperty("request_cache_expiry")
    public int request_cache_expiry;
    @JsonProperty("mal_id")
    public int mal_id;
    @JsonProperty("url")
    public String url;
    @JsonProperty("name")
    public String name;
    @JsonProperty("name_kanji")
    public String name_kanji;
    @JsonProperty("nicknames")
    public ArrayList<String> nicknames;
    @JsonProperty("about")
    public String about;
    @JsonProperty("member_favorites")
    public int member_favorites;
    @JsonProperty("image_url")
    public String image_url;
    @JsonProperty("animeography")
    public ArrayList<Animeography> animeography;
    @JsonProperty("mangaography")
    public ArrayList<Mangaography> mangaography;
    @JsonProperty("voice_actors")
    public ArrayList<Voice_actors> voice_actors;

    /**
     * Gets pictures
     *
     * @return pictures
     * @throws IOException    IOException
     * @throws ParseException ParseException
     */
    @JsonProperty
    public CompletableFuture<Pictures> getPictures() throws IOException, ParseException {
        return retrieve(Pictures.class, baseURL + "/character/" + mal_id + "/pictures");
    }

    @Override
    public String toString() {
        return "Character{" +
                ",mal_id=" + mal_id +
                ",\n url='" + url + '\'' +
                ",\n name='" + name + '\'' +
                ",\n name_kanji='" + name_kanji + '\'' +
                ",\n nicknames=" + nicknames.toString() +
                ",\n about='" + about + '\'' +
                ",\n member_favorites=" + member_favorites +
                ",\n image_url='" + image_url + '\'' +
                ",\n animeography=" + animeography +
                ",\n mangaography=" + mangaography +
                ",\n voice_actors=" + voice_actors +
                '}';
    }
}

File: src/main/java/com/github/doomsdayrs/jikan4java/data/model/main/character/Character.kt
package com.github.doomsdayrs.jikan4java.data.model.main.character

import com.github.doomsdayrs.jikan4java.common.JIKAN_URL
import com.github.doomsdayrs.jikan4java.core.JikanResult
import com.github.doomsdayrs.jikan4java.core.Retriever
import com.github.doomsdayrs.jikan4java.data.base.values.MyAnimeListID
import com.github.doomsdayrs.jikan4java.data.base.values.MyAnimeListImageURL
import com.github.doomsdayrs.jikan4java.data.base.values.MyAnimeListName
import com.github.doomsdayrs.jikan4java.data.base.values.MyAnimeListURL
import com.github.doomsdayrs.jikan4java.data.base.endpoint.MyAnimeListPicturesEndPoint
import com.github.doomsdayrs.jikan4java.data.base.endpoint.MyAnimeListSelfType
import com.github.doomsdayrs.jikan4java.data.base.endpoint.direct.MyAnimeListDirectPicturesEndPoint
import com.github.doomsdayrs.jikan4java.data.model.support.RequestHashing
import com.github.doomsdayrs.jikan4java.data.model.support.VoiceActors
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.*
import java.util.concurrent.CompletableFuture

/*
 * This file is part of Jikan4java.
 *
 * Jikan4java is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Jikan4java is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Jikan4java.  If not, see <https://www.gnu.org/licenses/>.
 * ====================================================================
 * Jikan4java
 * 28 / October / 2018
 *
 * @author github.com/doomsdayrs
 */
@Serializable
data class Character(
	@SerialName("request_hash") override val requestHash: String = "",
	@SerialName("request_cached") override val requestCached: Boolean = false,
	@SerialName("request_cache_expiry") override val requestCacheExpiry: Int = 0,
	@SerialName("mal_id") override val malID: Int = 0,
	@SerialName("url") override val url: String,
	@SerialName("name") override val name: String,
	@SerialName("name_kanji") val name_kanji: String = "",
	@SerialName("nicknames") val nicknames: ArrayList<String> = arrayListOf(),
	@SerialName("about") val about: String = "",
	@SerialName("member_favorites") val member_favorites: Int = 0,
	@SerialName("image_url") override val imageURL: String = "",
	@SerialName("animeography") val animeography: ArrayList<Animeography>,
	@SerialName("mangaography") val mangaography: ArrayList<Mangaography>,
	@SerialName("voice_actors") val voice_actors: ArrayList<VoiceActors>
) : RequestHashing, MyAnimeListID, MyAnimeListURL, MyAnimeListName,
	MyAnimeListImageURL, MyAnimeListPicturesEndPoint {
	override val urlPoint: String by lazy { "character" }

	companion object : MyAnimeListSelfType<Character>,
		MyAnimeListDirectPicturesEndPoint {
		override val urlPoint: String by lazy { "character" }

		override fun getByID(
			retriever: Retriever,
			id: Int
		): CompletableFuture<JikanResult<Character>> =
			retriever("$JIKAN_URL/$urlPoint/$id")
	}
}